src/libraries/System.Private.CoreLib/src/System/Random.ImplBase.cs

```diff
@@ -0,0 +1,59 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Numerics;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    public partial class Random
+    {
+        /// <summary>Base type for all generator implementations that plug into the base Random.</summary>
+        internal abstract class ImplBase
+        {
+            public abstract double Sample();
+
+            public abstract int Next();
+
+            public abstract int Next(int maxValue);
+
+            public abstract int Next(int minValue, int maxValue);
+
+            public abstract long NextInt64();
+
+            public abstract long NextInt64(long maxValue);
+
+            public abstract long NextInt64(long minValue, long maxValue);
+
+            public abstract float NextSingle();
+
+            public abstract double NextDouble();
+
+            public abstract void NextBytes(byte[] buffer);
+
+            public abstract void NextBytes(Span<byte> buffer);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)] // hot path, only a handful of callers, not otherwise inlined, and intrinsics result in a small amount of asm
+            protected static int Log2Ceiling(uint value)
```

Comment:
Should we place this as an internal method on `BitOperations`?

IMO, it would be nice if we could consistently place and look for such helpers there as it helps give an idea of usage, a central place for improvements, and an idea if we are possibly missing other public surface area.

Reply:
Sure.

Reply:
Just need to be careful about how the zero input is handled. I suggest changing the != 1 to > 1 if you move it to BitOperations, like so:

```
            public static int Log2Ceiling(uint value)
            {
                int result = BitOperations.Log2(value);
                if (BitOperations.PopCount(value) > 1)
                {
                    result++;
                }
                return result;
            }
```

I assume the JITter does its magic, and thus the above is no different in reality to, e.g.

```
            int exp = BitOperations.Log2(x);
            return BitOperations.PopCount(x) > 1 ? exp + 1 : exp;
```

Also need to ponder possible overloads for int, uint, ulong. I presume they would all return int though(?)

Reply:
Right, if we ever made it public, those things would need to be considered. But until then, it still gives a central location for anyone adding other simple helpers in the future to look.

Reply:
> I assume the JITter does its magic, and thus the above is no different in reality to

Not 100%.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGjWIBmJqQYBhBgG8aDJUwnMkDXgDsMDADIQA5qWkxeAGy0HyACg5adAN2xmOMAJSLlC6st8btDLC4HGY6ALwMAEK8GADyAA482Bi8EJq4LPpG1k4u7sI+fkq8AGYM1tFxiVDJqeksAAoQ8dIQHNo5zq5uDAB8DOQehUXeRUVBIRgA1FMFYwwAvp5FxADsgTDBoXPKS9TLKkzk6vZ6hsamFprZdgG53QejY6cwCPEMEZUJSSlpGVmkTp5Nw7FbrL7VWp/RrNVrtDBA7p9AYMAD8DFe7ymKJAGLeoMWNAWQA

> Just need to be careful about how the zero input is handled

I'm going to leave it as is. It's internal.  If we add consumers that might need to worry about 0 or if it's made public, we can change it then to whatever behavior is deemed best.

Reply:
Footnote / curiosity. It seemed to me that we might be able to eliminate the conditional branch in Log2Ceiling. I looked into it, and it is possible, but the performance is pretty awful compared to what we have (about 10x slower in synthetic benchmarks on my core i7):

        public static int Log2Ceiling(uint x)
        {
            int exp = BitOperations.Log2(x);

            if (BitOperations.PopCount(x) > 1)
            {
                exp++;
            }

            return exp;
        }

        public static int Log2Ceiling2(uint x)
        {
            int lzcnt = BitOperations.LeadingZeroCount(x | 1);

            // Clear high bit of x.
            x &= ~(1u << (31 ^ lzcnt));

            // Add one to final result if any of the remaining bits are set.
            // Popcount method.
            return (31 ^ lzcnt) + BitOperations.PopCount((uint)BitOperations.PopCount((uint)BitOperations.PopCount((uint)BitOperations.PopCount(x))));
        }

        public static int Log2Ceiling3(uint x)
        {
            int lzcnt = BitOperations.LeadingZeroCount(x | 1u);

            // Clear high bit of x.
            x &= ~(1u << (31 ^ lzcnt));

            // Add one to final result if any of the remaining bits are set.
            // Trailing zero shift method.

            // Of the remaining bits, find position of lowest set bit (if any).
            int tzcnt = BitOperations.TrailingZeroCount(x);

            // Shift lowest set bit to position zero, and clear all other bits.
            // Add to 31 - lzcnt.
            return (31 ^ lzcnt) + (int)((x >> tzcnt) & 0x1u);
        }


All of the extra bit manipulation ops just aren't worth it. Was worth a shot I guess :|

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgDkBXfGKASzFwBuGjWIBmJqQYBhBgG8aDJcoaKVy8U3JIGvAHYYGAGQgBzUtJi8ANvtMAKDvsMIAlGvXyPn5QHpfxmak9m4MprwAbjC4DBgAFjAMsACOTrAAJgwADhAA7jwMEABmsbkQaAxxeTBRULFxvDGNugYwpgUm5vau9TwwRdCJ2HoAngxFUNhgGLwQetjW3j4M/tnYUIb6vUnRHNabMRhQHHpg2Bgw6RW8LDAsFfGJsLh7hvjYY8CJ5LEQEAzWPIsBgAFX+kHwWRgelw50S+QY2HSmR+vBKCCWPlWzT0EEMwwYMAQU0MOXydWKpQgLExnmchIQWQYAF4GAAhXgYADyUMmMzmuBYnWCbmE1Fp6lWACUYBgOFA9Ax7ESmQBqBjkHpohgIXQxXF6OAALx45URekyuPxiqJJOy1QpJQwZRp1GWKm19g53N551mMJYAAUIFlpBAThgQj0AHwa9xu91KBQJxMqFWq1Vi1MMAC+IhT7uIAHYGVksz48+KCypNMwdPThZYbHZgk4DDr48tk9n6dYjWB26zvTyeH6BUKYEi7AAtU1hiMhBiEONiiUqVbSayTuoNUxxBjATmFdGu7O6gBkrIAfvZyBwGAAeB9KsQ/AB6AP7BlcrlX1axAQAILIoUeiJBg/xFPoCw7C8+y6CUwxjJSjw7O8+h2AenIxOsiS4LKp6pqswZZJAEYMNw8QQOkhGJsWL7vp+A4YD06rDr6/IBiR84GPYjjOK47GjpxgrceGvH8d+Ql8v6okhjxkaSSx0ljlx8niZGbg/r+EqVhKtbaC0hiNlYth6KYYhKR2ErdqmvZfoYQ6ciOMnjkYk7pDOc4aYuy53jpVbZhuW7rJUvB7lhhiUggtHuhe163veT4MQwH59sxP5/kFQEgXM4GQdB1iwa8CHmshToJGh2AYeZkU4bADD4RgsXLKsIKTM2tUmlA/y4A0RRvLKVQ0fm2UMFyFVPDA6F6Jhh4YLgFRQRa9q4Jy/rHgC1S4IYTWRUq2pIa4LU+PSGAOSy7LORxsksO11VmaYs49QpUZZURAQAMr9YYgL5DtjWyvtEGretcwMN1ZrDJkYAhXUCxFXiCR1PNgprn4OWZCDr4MHATEGCdnj0fYONpQ5rEHd+fG6tGsbnRlDDngwdAIP55aeJWOZAA